[BoundsSafety][Sema] Allow counted_by and counted_by_or_null on pointers where the pointee type is incomplete but potentially completable

delcypher opened the pull request:
Previously using the `counted_by` or `counted_by_or_null` attribute on a pointer with an incomplete pointee type was forbidden. Unfortunately this prevented a situation like the following from being allowed.

Header file:

```
struct EltTy; // Incomplete type
struct Buffer {
  size_t count;
  struct EltTy* __counted_by(count) buffer; // error before this patch
};
```

Implementation file:

```
struct EltTy {
  // definition
};

void allocBuffer(struct Buffer* b) {
  b->buffer = malloc(sizeof(EltTy)* b->count);
}
```

To allow code like the above but still enforce that the pointee type size is known in locations where `-fbounds-safety` needs to emit bounds checks the following scheme is used.

* For incomplete pointee types that can never be completed (e.g. `void`) these are treated as error where the attribute is written (just like before this patch).
* For incomplete pointee types that might be completable later on (struct, union, and enum forward declarations) in the translation unit, writing the attribute on the incomplete pointee type is allowed on the FieldDecl declaration but "uses" of the declared pointer are forbidden if at the point of "use" the pointee type is still incomplete.

For this patch a "use" of a FieldDecl covers:

* Explicit and Implicit initialization (note see **Tentative Definition Initialization** for an exception to this)
* Assignment
* Conversion to an lvalue (e.g. for use in an expression)

In Apple's internal fork of Clang the `counted_by` and `counted_by_or_null` attribute are allowed in many more contexts. That isn't the case for upstream Clang so the "use" checks for the attribute on VarDecl, ParamVarDecl, and function return type have been omitted from this patch because they can't be tested. However, the `BoundsSafetyCheckAssignmentToCountAttrPtrWithIncompletePointeeTy` and `BoundsSafetyCheckUseOfCountAttrPtrWithIncompletePointeeTy` functions retain the ability to emit diagnostics for these other contexts to avoid unnecessary divergence between upstream Clang and Apple's internal fork. Support for checking "uses" will be upstreamed when upstream Clang allows the `counted_by` and `counted_by_or_null` attribute in additional contexts.

This patch has a few limitations:

** 1. Tentative Defition Initialization **

This patch currently allows something like:

```
struct IncompleteTy;
struct Buffer {
  int count;
  struct IncompleteTy* __counted_by(count) buf;
};

// Tentative definition
struct Buffer GlobalBuf;
```

The Tentative definition in this example becomes an actual definition whose initialization **should be checked** but it currently isn't. Addressing this problem will be done in a subseqent patch.

** 2. When the incomplete pointee type is a typedef diagnostics are slightly misleading **

For this situation:

```

struct IncompleteTy;
typedef struct IncompleteTy Incomplete_t;

struct Buffer {
  int count;
  struct IncompleteTy* __counted_by(count) buf;
};

void use(struct Buffer b) {
  b.buf = 0x0;
}
```

This code emits `note: forward declaration of 'Incomplete_t' (aka 'struct IncompleteTy')` but the location is on the `struct IncompleteTy;` forward declaration.  This is misleading because `Incomplete_t` isn't actually forward declared there (instead the underlying type is). This could be resolved by additional diagnostics that walk the chain of typedefs and explain each step of the walk. However, that would be very verbose and didn't seem like a direction worth pursuing.

rdar://133600117

delcypher committed:
[BoundsSafety][Sema] Allow counted_by and counted_by_or_null on pointers where the pointee type is incomplete but potentially completable

Previously using the `counted_by` or `counted_by_or_null` attribute on a
pointer with an incomplete pointee type was forbidden. Unfortunately
this prevented a situation like the following from being allowed.

Header file:

```
struct EltTy; // Incomplete type
struct Buffer {
  size_t count;
  struct EltTy* __counted_by(count) buffer; // error before this patch
};
```

Implementation file:

```
struct EltTy {
  // definition
};

void allocBuffer(struct Buffer* b) {
  b->buffer = malloc(sizeof(EltTy)* b->count);
}
```

To allow code like the above but still enforce that the pointee
type size is known in locations where `-fbounds-safety` needs to
emit bounds checks the following scheme is used.

* For incomplete pointee types that can never be completed (e.g. `void`)
  these are treated as error where the attribute is written (just like
  before this patch).
* For incomplete pointee types that might be completable later on
  (struct, union, and enum forward declarations)
  in the translation unit, writing the attribute on the incomplete
  pointee type is allowed on the FieldDecl declaration but "uses" of the
  declared pointer are forbidden if at the point of "use" the pointee
  type is still incomplete.

For this patch a "use" of a FieldDecl covers:

* Explicit and Implicit initialization (note see **Tentative Definition
  Initialization** for an exception to this)
* Assignment
* Conversion to an lvalue (e.g. for use in an expression)

In the swift lang fork of Clang the `counted_by` and
`counted_by_or_null` attribute are allowed in many more contexts. That
isn't the case for upstream Clang so the "use" checks for the attribute
on VarDecl, ParamVarDecl, and function return type have been omitted
from this patch because they can't be tested. However, the
`BoundsSafetyCheckAssignmentToCountAttrPtrWithIncompletePointeeTy` and
`BoundsSafetyCheckUseOfCountAttrPtrWithIncompletePointeeTy` functions
retain the ability to emit diagnostics for these other contexts to avoid
unnecessary divergence between upstream Clang and Apple's internal fork.
Support for checking "uses" will be upstreamed when upstream Clang
allows the `counted_by` and `counted_by_or_null` attribute in additional
contexts.

This patch has a few limitations:

** 1. Tentative Defition Initialization **

This patch currently allows something like:

```
struct IncompleteTy;
struct Buffer {
  int count;
  struct IncompleteTy* __counted_by(count) buf;
};

// Tentative definition
struct Buffer GlobalBuf;
```

The Tentative definition in this example becomes an actual definition
whose initialization **should be checked** but it currently isn't.
Addressing this problem will be done in a subseqent patch.

** 2. When the incomplete pointee type is a typedef diagnostics are slightly misleading **

For this situation:

```

struct IncompleteTy;
typedef struct IncompleteTy Incomplete_t;

struct Buffer {
  int count;
  struct IncompleteTy* __counted_by(count) buf;
};

void use(struct Buffer b) {
  b.buf = 0x0;
}
```

This code emits `note: forward declaration of 'Incomplete_t' (aka
'struct IncompleteTy')` but the location is on the `struct
IncompleteTy;` forward declaration.  This is misleading because
`Incomplete_t` isn't actually forward declared there (instead the
underlying type is). This could be resolved by additional diagnostics
that walk the chain of typedefs and explain each step of the walk.
However, that would be very verbose and didn't seem like a direction
worth pursuing.

rdar://133600117

delcypher commented:
@bwendling @kees Apologies for taking so long to get this change upstream. We spent quite a lot of time experimenting with different implementations that ended up not working out. Our internal version of this patch is larger because `counted_by` is allowed in many more contexts than it is in upstream Clang so some of that has been omitted from this patch. I can upstream the rest once we have support for `counted_by` is more contexts.

delcypher commented on clang/include/clang/Sema/Sema.h:
```diff
@@ -2099,6 +2099,70 @@ class Sema final : public SemaBase {
 
+  // AssignmentAction - This is used by all the assignment diagnostic functions
+  // to represent what is actually causing the operation
+  enum AssignmentAction {
```
@Endilll Any opinions on where `AssignmentAction` should go? Unfortunately I had to move it because the enum needs to be defined before I can use it as a parameter type.

This wasn't a problem in the original version of my patch but when I rebased on latest `next` it appears a recent refactor (9f1dc01e0ab4f91c4052a712ce590d2e86b81dc3) moved where the `BoundsSafety*` methods are declared so that they are now **before** where `AssignmentAction`'s declaration. Previously the methods were declared after `AssignmentActions`'s declaration.

Endilll replied:
Ideally you should turn it into scoped enum and move it outside of `Sema` definition in a separate NFC patch, then update this PR with those changes.

delcypher replied:
@Endilll I created #106453 which I think is what you're asking for.

Sirraide commented on clang/lib/Sema/SemaBoundsSafety.cpp:
```diff
@@ -186,4 +218,370 @@ bool Sema::CheckCountedByAttrOnField(FieldDecl *FD, Expr *E, bool CountInBytes,
   return false;
 }
 
+SourceRange Sema::BoundsSafetySourceRangeFor(const CountAttributedType *CATy) {
```
This entire function seems like a very bad idea to me: I don’t think Sema should ever have to fiddle with source locations in terms of character offsets or anything like that. This seems extremely brittle in case a user writes something like e.g. `__counted_by (  count  )` or uses macros (which you mention in a comment doesn’t work).

I think you should be able to get a hold of the corresponding `CountAttributedTypeLoc` somehow (e.g. via the `FieldDecl`, which has a `TypeSourceInfo*`, which you should be able to get the type loc from; you might have to do this a few levels up the call stack before we ever get here) and then use its source location and range.

delcypher replied:
Yeah this isn't great. The whole function is essentially a workaround for the fact we don't have a good way to get `CountAttributedTypeLoc` from a `CountAttributedType`.

Given that I don't want to solve this problem right now I think simplest thing to do is just use the SourceRange from the count expression (which is the fallback for this function anyway). It does mean the SourceRange is slightly wrong (doesn't actually include the attribute text) but this isn't a deal breaker.

I can then tackle fixing this as part of the work  to fix the issue mentioned in this comment: https://github.com/llvm/llvm-project/pull/106321#discussion_r1735207309

Does that sound reasonable to you?

erichkeane replied:
I agree with Sirraide here, but would have to spend some time contemplating the implications of your proposal.

delcypher replied:
I've dropped the hacky implementation in favor of just using the SourceRange attached to the count expression. I've filed #113582 to track resolving this properly.

Sirraide commented on clang/lib/Sema/SemaBoundsSafety.cpp:
```diff
@@ -186,4 +218,370 @@ bool Sema::CheckCountedByAttrOnField(FieldDecl *FD, Expr *E, bool CountInBytes,
+///        determined.
+///
+/// \returns True iff no diagnostic where emitted, false otherwise.
+static bool BoundsSafetyCheckAssignmentToCountAttrPtrWithIncompletePointeeTy(
```
`static` functions in this file don’t really need to be prefixed with `BoundsSafety`.

Sirraide commented on clang/lib/Sema/SemaBoundsSafety.cpp:
```diff
@@ -186,4 +218,370 @@ bool Sema::CheckCountedByAttrOnField(FieldDecl *FD, Expr *E, bool CountInBytes,
+/// \returns True iff no diagnostic where emitted, false otherwise.
+static bool BoundsSafetyCheckAssignmentToCountAttrPtrWithIncompletePointeeTy(
+    Sema &S, QualType LHSTy, Expr *RHSExpr, Sema::AssignmentAction Action,
+    SourceLocation Loc, std::function<std::string()> ComputeAssignee) {
```
```suggestion
    SourceLocation Loc, llvm::function_ref<std::string()> ComputeAssignee) {
```

Sirraide commented on clang/lib/Sema/SemaBoundsSafety.cpp:
```diff
@@ -186,4 +218,370 @@ bool Sema::CheckCountedByAttrOnField(FieldDecl *FD, Expr *E, bool CountInBytes,
+  std::string Assignee = ComputeAssignee ? ComputeAssignee() : "";
+  {
+    auto D = S.Diag(Loc, diag::err_counted_by_on_incomplete_type_on_assign)
+             << /*0*/ (int)Action << /*1*/ Assignee
+             << /*2*/ (Assignee.size() > 0)
+             << /*3*/ isa<ImplicitValueInitExpr>(RHSExpr) << /*4*/ LHSTy
+             << /*5*/ CATy->GetAttributeName(/*WithMacroPrefix=*/true)
+             << /*6*/ PointeeTy << /*7*/ CATy->isOrNull();
```
```suggestion
             << (int)Action << Assignee
             << (Assignee.size() > 0)
             << isa<ImplicitValueInitExpr>(RHSExpr) << LHSTy
             << CATy->GetAttributeName(/*WithMacroPrefix=*/true)
             << PointeeTy << CATy->isOrNull();
```
I don’t think the numbers really add much of value here (it’s not that hard to count e.g. to the 5th parameter if you want to figure out what that’s being used for ;Þ)

Sirraide commented on clang/lib/Sema/SemaBoundsSafety.cpp:
```diff
@@ -186,4 +218,370 @@ bool Sema::CheckCountedByAttrOnField(FieldDecl *FD, Expr *E, bool CountInBytes,
+
+bool Sema::BoundsSafetyCheckAssignmentToCountAttrPtr(
+    QualType LHSTy, Expr *RHSExpr, Sema::AssignmentAction Action,
+    SourceLocation Loc, std::function<std::string()> ComputeAssignee) {
```
```suggestion
    SourceLocation Loc, llvm::function_ref<std::string()> ComputeAssignee) {
```

Sirraide commented on clang/lib/Sema/SemaBoundsSafety.cpp:
```diff
@@ -186,4 +218,370 @@ bool Sema::CheckCountedByAttrOnField(FieldDecl *FD, Expr *E, bool CountInBytes,
+  return ChecksPassed;
+}
+
+static bool BoundsSafetyCheckUseOfCountAttrPtrWithIncompletePointeeTy(Sema &S,
```
```suggestion
static bool CheckUseOfCountAttrPtrWithIncompletePointeeTy(Sema &S,
```

Sirraide replied:
Actually, can we just inline this? I think it’s only called in one place?

delcypher replied:
My original thinking behind this was that `Sema::BoundsSafetyCheckUseOfCountAttrPtr()` would grow other time to check more things which is why things were separate out. However, for now it's fine to inline it and I can outline again if needed.

Sirraide commented on clang/lib/Sema/SemaBoundsSafety.cpp:
```diff
@@ -186,4 +218,370 @@ bool Sema::CheckCountedByAttrOnField(FieldDecl *FD, Expr *E, bool CountInBytes,
+  }
+  assert(UseStr.size() > 0);
+
+  S.Diag(E->getBeginLoc(), diag::err_counted_by_on_incomplete_type_on_use)
+      << /*0*/ SelectExprKind << /*1*/ UseStr << /*2*/ T << /*3*/ PointeeTy
+      << /*4*/ CATy->GetAttributeName(/*WithMacroPrefix=*/true)
+      << /*5*/ CATy->isOrNull() << E->getSourceRange();
```
```suggestion
  S.Diag(E->getBeginLoc(), diag::err_counted_by_on_incomplete_type_on_use)
      << SelectExprKind << UseStr << T << PointeeTy
      << CATy->GetAttributeName(/*WithMacroPrefix=*/true)
      << CATy->isOrNull() << E->getSourceRange();
```

delcypher replied:
I'll drop the comments.

Sirraide commented on clang/lib/Sema/SemaInit.cpp:
```diff
@@ -8192,6 +8192,14 @@ ExprResult InitializationSequence::Perform(Sema &S,
             Kind.getRange().getEnd());
       } else {
         CurInit = new (S.Context) ImplicitValueInitExpr(Step->Type);
+        // Note the return value isn't used to return early
+        // to preserve the AST as best as possible even though an error
+        // might have occurred. For struct initialization it also allows
+        // all field assignments to be checked rather than bailing on the
+        // first error.
+        S.BoundsSafetyCheckInitialization(Entity, Kind, Sema::AA_Initializing,
```
Not sure we really need this comment as that much is pretty much implied by discarding the return value (this comment is also duplicated somewhere else, which isn’t exactly ideal). I think we can just remove this?

delcypher replied:
The comment is explaining **why** we aren't checking the return value. If you remove it isn't obvious at all why its being ignored.

Sirraide replied:
(just looked at the other comment; I’d say just apply whatever consensus we arrive at for that other comment here too ;Þ)

Sirraide replied:
Actually, wait, I just noticed that there are several places we do this. I don’t remember off the top of my head, but is there any place where we actually *do* check for the error? If we always ignore it everywhere, the function should probably just return `void`, and the comment as to why we ‘ignore’ the error should be on the function instead.

Sirraide replied:
> The return value isn't always ignored in our fork.

Just saw this comment; I think it’s fine to keep the return value then, but I’m not sure we want to copy-paste the same comment across like 4 different calls to this function. Having it in one location where it’s maybe the least self-evident seems fine, but I think we generally want to eschew code duplication, even if it’s just a comment and not code.

Some places have (or at one point had) a cast to `void`; that I think we can remove everywhere.

Sirraide commented on clang/lib/Sema/SemaInit.cpp:
```diff
@@ -8238,6 +8246,15 @@ ExprResult InitializationSequence::Perform(Sema &S,
         }
       }
 
+      // Note the return value isn't used to return early so that additional
+      // diagnostics can be emitted and to preserve the AST as best as possible
+      // even though an error might have occurred. For struct initialization it
+      // also allows all field assignments to be checked rather than bailing on
+      // the first error.
+      (void)S.BoundsSafetyCheckInitialization(
```
Same here. The cast to `(void)` also isn’t necessary I’d say.

delcypher replied:
The comment is explaining why we aren't checking the return value. If you remove it isn't obvious at all why its being ignored.

delcypher replied:
Marking this as resolved (even though it technically isn't) for ease of reviewing because this is being discussed under https://github.com/llvm/llvm-project/pull/106321#discussion_r1816031253

Sirraide commented on clang/lib/Sema/SemaInit.cpp:
```diff
@@ -8192,6 +8192,14 @@ ExprResult InitializationSequence::Perform(Sema &S,
+        // all field assignments to be checked rather than bailing on the
+        // first error.
+        S.BoundsSafetyCheckInitialization(Entity, Kind, Sema::AA_Initializing,
+                                          /*LHSType=*/Step->Type,
+                                          /*RHSExpr=*/CurInit.get());
```
Also, I think we usually only use `/*name=*/` comments for bool literals and `{}`.

delcypher replied:
This is here for clarify because it really isn't obvious that `Step->Type` means the LHS type (similarly for the other parameters).

The style guide says:

https://llvm.org/docs/CodingStandards.html#comment-formatting

---

> When documenting the significance of constants used as actual parameters in a call. This is most helpful for bool parameters, or passing 0 or nullptr. The comment should contain the parameter name, which ought to be meaningful. For example, it’s not clear what the parameter means in this call:

```
Object.emitName(nullptr);
```

> An in-line C-style comment makes the intent obvious:

```
Object.emitName(/*Prefix=*/nullptr);
```

---

So you're right that it's encouraged for constant arguments which isn't the case here. I'll drop the comments.

Sirraide reviewed (changes requested):
I’m pretty sure there are some fundamental problems with the current approach wrt how getting source locations for the attribute is being handled. 

The general approach for the rest of it seems fine from a cursory look, though. :)

delcypher commented on clang/include/clang/AST/Type.h:
```diff
@@ -2440,6 +2440,26 @@ class alignas(TypeAlignment) Type : public ExtQualsTypeCommonBase {
+  /// // This decl how has complete type 'char[5]'.
+  /// char foo[5]; // foo has a complete type
+  /// \endcode
+  bool isIncompletableIncompleteType() const;
```
This is a method and not inlined at the use site for multiple reasons:

1. This method is **also** called from a different location in our internal fork of Clang. The reason for this is because the Sema code for handling the `counted_by` in upstream Clang and our internal fork are completely different. We plan to eventually merge the two different Sema codes and open source it but won't be doing that until upstream Clang allows `counted_by` in all the contexts that it is allowed in our internal fork.

2. "isIncompletableIncompleteType" is a property of the type so the natural place for this code to live is in the `Type` class.

erichkeane replied:
That suggestion makes sense, the original name/current name does not really tell me anything.

delcypher commented on clang/lib/AST/Type.cpp:
```diff
@@ -2438,6 +2438,22 @@ bool Type::isIncompleteType(NamedDecl **Def) const {
   }
 }
 
+bool Type::isIncompletableIncompleteType() const {
+  if (!isIncompleteType())
+    return false;
+
+  // Forward declarations of structs, classes, enums, and unions could be later
+  // completed in a compilation unit by providing a type definition.
+  if (isStructureOrClassType() || isEnumeralType() || isUnionType())
+    return false;
+
+  // Other types are incompletable.
+  //
+  // E.g. `char[]` and `void`. The type is incomplete and no future
+  // type declarations can make the type complete.
+  return true;
+}
+
```
This implementation is more concise (and is probably slightly faster) but it is much less readable. I could potentially fix that by using the conditions if your patch but not writing them like this.

```
if (!isCompleteType())
  return false;

// Forward declarations of structs, classes, enums, and unions could be later
// completed in a compilation unit by providing a type definition.
TagDecl *TD = getAsTagDecl();
if (TD)
  return false;

// Other types are incompletable.
//
// E.g. `char[]` and `void`. The type is incomplete and no future
// type declarations can make the type complete.
return true;
```

Why is the `!TD->isCompleteDefinition()` necessary? If `isIncompleteType()` returned true then that would imply that `!TD->isCompleteDefinition()` is true, right? Or am I missing something?

delcypher replied:
As noted in another comment I'm not going to inline this function.

delcypher commented on clang/lib/AST/Type.cpp:
```diff
@@ -3862,6 +3878,27 @@ CountAttributedType::CountAttributedType(
     DeclSlot[i] = CoupledDecls[i];
 }
 
+StringRef CountAttributedType::GetAttributeName(bool WithMacroPrefix) const {
```
>  but the way this works for AttributedTypes is that the Attr* is stored in the AttributedTypeLoc

How does that help in the general case? If all you have is the `AttributedType` then you can't get a `AttributedTypeLoc` from that. I suspect that's by design because I guess an attributed type could be spelled out in multiple locations in a source file.

> just store it in and then get it from the CountAttributedTypeLoc instead.

Do mean store a `CountAttributedType` pointer in `CountAttributedTypeLoc`?

In general using TypeLocs here seems like it's going to be complicated.

My naive understanding of TypeLocs is that you have to fetch those from declarations using `getTypeSourceInfo()->getTypeLoc()`, is that right or am I missing something? This means we'd have the functions that issue diagnostics traverse expressions to find the relevant decls to find the TypeLocs. That might be ok for upstream Clang where the attribute is only allowed on `FieldDecl`s but it's more complicated for our internal fork because the Attribute is allowed in many more places.

I'm also not sure how that is supposed to work with casts.

```
void test(void) {
  int count = 5;
  // How would you get a TypeLoc for a the C-style cast below? The cast isn't a decl.
  char* __counted_by(count) ptr = (char* __counted_by(count)) malloc(count);
}
```

If I've understood your suggestion correctly then what you're suggesting does sound like an improvement (that would also solve the issue this patch has with trying to get the SourceRange for the attribute) but I think the best way to proceed here would be to file an issue about this problem and refer to it in a comment in `CountAttributedType::GetAttributeName` saying that this should be implemented in a different way. That way landing this patch isn't blocked on this issue. And once I've landed this PR then I can start looking into how to fix that issue that is compatible with upstream and our internal fork and then upstream that.

Does that seem reasonable to you?

delcypher replied:
I've filed #113585 to track fixing the properly. I don't think we can fix this until we also fix #113582

erichkeane commented on clang/include/clang/AST/Type.h:
```diff
@@ -3343,6 +3363,8 @@ class CountAttributedType final
     return T->getTypeClass() == CountAttributed;
   }
+
+  StringRef GetAttributeName(bool WithMacroPrefix) const;
```
```suggestion
  StringRef getAttributeName(bool WithMacroPrefix) const;
```

Also a little concerning how specialized this is, rather than being part of how this type is printed.

delcypher replied:
I filed #113585 to track fixing this properly. I don't see a way of fixing this until we have a good way to get the TypeLoc for the attribute (#113582).

erichkeane commented on clang/lib/Sema/SemaBoundsSafety.cpp:
```diff
@@ -186,4 +218,370 @@ bool Sema::CheckCountedByAttrOnField(FieldDecl *FD, Expr *E, bool CountInBytes,
+  // unique means the SourceLocation of the `counted_by` expression can be used
+  // to find where the attribute was written.
+
+  auto Fallback = CATy->getCountExpr()->getSourceRange();
```
I don't think 'auto' is supposed to be used in most of the uses in the function here.

delcypher replied:
This method's implementation has been completely changed so this isn't relevant anymore.

erichkeane commented on clang/lib/Sema/SemaExpr.cpp:
```diff
@@ -13559,6 +13562,27 @@ QualType Sema::CheckAssignmentOperands(Expr *LHSExpr, ExprResult &RHS,
+    // Even if this check fails don't return early to allow the best
+    // possible error recovery and to allow any subsequent diagnostics to
+    // work.
+    (void)BoundsSafetyCheckAssignmentToCountAttrPtr(
```
Why the void cast here?

delcypher replied:
It was to suppress any compiler that might want to warn about the return value being ignored.

delcypher replied:
I'll remove them

AaronBallman replied:
Instead of removing the `void` cast, should the return type from that function be changed to `void` given that the only call to it ignores the value?

delcypher replied:
The return value isn't always ignored in our fork. See [here](https://github.com/swiftlang/llvm-project/blob/b0bbd7e8a003990aff9ce373bc6b83407630b5f5/clang/lib/Sema/SemaBoundsSafety.cpp#L645)

delcypher replied:
@AaronBallman The return value isn't always ignored in our fork. See [here](https://github.com/swiftlang/llvm-project/blob/b0bbd7e8a003990aff9ce373bc6b83407630b5f5/clang/lib/Sema/SemaBoundsSafety.cpp#L645)

erichkeane reviewed (commented):
I did a pass, I think @Sirraide is on the right direction in his reviews here, but the general direction of this patch is also completely acceptable.

delcypher commented:
@Sirraide @erichkeane  Sorry for the delay in addressing your feedback. This PR is ready for another round of reviews. Note I'm aware there are some code formatting problems. I'll fix them when I squash locally. I've tried to keep all my tweaks as separate commits so it's easier to review.

Sirraide commented on clang/test/Sema/attr-counted-by-struct-ptrs-completable-incomplete-pointee.c:
```diff
@@ -0,0 +1,584 @@
+// =============================================================================
+
+// expected-note@+1 8{{forward declaration of 'IncompleteEnum_ty' (aka 'enum IncompleteEnumTy')}}
+enum IncompleteEnumTy; // expected-note 8{{forward declaration of 'enum IncompleteEnumTy'}}
```
Note that this is valid if we specify an explicit underlying type:
```c
enum s : unsigned;
enum s x; // Ok
```
I would expect that this already works as it should, but we should probably also have a test or two for that as well.

delcypher replied:
I've added some tests for this to `clang/test/Sema/attr-counted-by-struct-ptrs-completable-incomplete-pointee.c`

erichkeane commented on clang/lib/Sema/SemaBoundsSafety.cpp:
```diff
@@ -186,4 +218,216 @@ bool Sema::CheckCountedByAttrOnField(FieldDecl *FD, Expr *E, bool CountInBytes,
+      << CATy->getPointeeType();
+}
+
+static bool
+HasCountedByAttrOnIncompletePointee(QualType Ty, NamedDecl **ND,
+                                    const CountAttributedType **CATyOut,
+                                    QualType *PointeeTyOut) {
```
This seems like the right answer too IMO.

AaronBallman replied:
+1, three out params is a bit much.

delcypher replied:
Thanks for the suggestion. This is so much cleaner. I've made the change.

erichkeane commented on clang/lib/Sema/SemaBoundsSafety.cpp:
```diff
@@ -186,4 +218,216 @@ bool Sema::CheckCountedByAttrOnField(FieldDecl *FD, Expr *E, bool CountInBytes,
+  std::string Assignee = ComputeAssignee ? ComputeAssignee() : "";
+  {
+    auto D = S.Diag(Loc, diag::err_counted_by_on_incomplete_type_on_assign)
+             << (int)Action << Assignee << (Assignee.size() > 0)
```
```suggestion
             << static_cast<int>(Action) << Assignee << (Assignee.size() > 0)
```

erichkeane reviewed (commented):
@Sirraide 's comments are all relevant, and he's complained about everything I would have.  So when he's happy with this, I am too.

delcypher commented on clang/lib/Sema/SemaInit.cpp:
```diff
@@ -8207,6 +8207,14 @@ ExprResult InitializationSequence::Perform(Sema &S,
             Kind.getRange().getEnd());
       } else {
         CurInit = new (S.Context) ImplicitValueInitExpr(Step->Type);
+        // Note the return value isn't used to return early
+        // to preserve the AST as best as possible even though an error
+        // might have occurred. For struct initialization it also allows
+        // all field assignments to be checked rather than bailing on the
+        // first error.
```
It would be superfluous if not for the part about struct initialization. To me that is really not very obvious. When possible I think it's very useful to comment **why** the code is doing something. All too often I come across code where I understand what it's doing but I understand **why**.

So I would prefer to keep it. If you prefer I could only keep the bit about struct initialization.

Sirraide replied:
> It would be superfluous if not for the part about struct initialization. To me that is really not very obvious.

That’s fair. I might just be to used to this as a pattern. Commenting *why* we don’t just throw the entire thing out of the window here and still try to keep operating on it even if this fails is probably fine.

> I've changed the comments to just be.
> 
> ```
>         // Note the return value isn't used to return a ExprError() when
>         // initialization fails . For struct initialization allows all field
>         // assignments to be checked rather than bailing on the first error.
> ```

I’d honestly shorten it even more to just:
```
// Don't return here so we can still check field assingments.
```
because e.g. the fact that ‘the return value isn't used to return a ExprError() when initialization fails’ is apparent from the fact that... there is no `return` statement here. ;Þ

delcypher commented on clang/test/Sema/attr-counted-by-struct-ptrs-completable-incomplete-pointee.c:
```diff
@@ -0,0 +1,584 @@
+  ptr->buf_typedef = 0x0;
+
+
+  // ===========================================================================
+  // ## Make sure modifying the fields through other assignment operators is not
+  //    allowed
+  // ===========================================================================
```
> Because isn’t this just testing that we don’t allow pointer arithmetic on incomplete types

Yes... but it's specifically testing we are disallowing pointer arithmetic on pointers with the `__counted_by` attribute which is unlikely to be tested elsewhere.  Also given that the test is testing how how accesses to pointer fields with the `__counted_by` attribute are handled it seemed very natural to also test how pointer arithmetic is handled.

If your **really** don't like this I can remove it but I personally think we should keep this part of the test.

Sirraide replied:
> Yes... but it's specifically testing we are disallowing pointer arithmetic on pointers with the `__counted_by` attribute which is unlikely to be tested elsewhere.

It’s fine to have tests for that sure, but my point is that if you try to perform pointer arithmetic on an *incomplete* type, we’re going to reject that anyway, irrespective of whether the member in question has the `__counted_by` attribute or not (https://godbolt.org/z/Ws8d38z6K).

So if you want to add test for arithmetic on `__counted_by`-attributed types, the type should be complete for that.

delcypher replied:
>  but my point is that if you try to perform pointer arithmetic on an incomplete type, we’re going to reject that anyway, irrespective of whether the member in question has the __counted_by attribute or not

But we should test that... That is indeed what I expect but that expectation for the `__counted_by` attribute should be tested.

delcypher commented on clang/test/Sema/attr-counted-by-struct-ptrs-completable-incomplete-pointee.c:
```diff
@@ -0,0 +1,584 @@
+  explicit_desig_init.buf_typedef = 0x0;
+  // expected-error@+1{{cannot use 'explicit_desig_init.buf' with type 'struct IncompleteTy * __counted_by(count)' (aka 'struct IncompleteTy *') because the pointee type 'struct IncompleteTy' is incomplete and the '__counted_by' attribute requires the pointee type be complete in this context; consider providing a complete definition for 'struct IncompleteTy' or using the '__sized_by' attribute}}
+  void *tmp = explicit_desig_init.buf;
+  // expected-error@+1{{cannot use 'explicit_desig_init.buf_typedef' with type 'Incomplete_t * __counted_by(count)' (aka 'struct IncompleteTy *') because the pointee type 'Incomplete_t' (aka 'struct IncompleteTy') is incomplete and the '__counted_by' attribute requires the pointee type be complete in this context; consider providing a complete definition for 'Incomplete_t' or using the '__sized_b ...
```
Ideally I think the all the information necessary for understanding an error should be contained within the `error` diagnostic itself and additional useful (but not strictly necessary) information can (but doesn't necessarily have to be) put into notes.

What is necessary is a little subjective but...

> ‘Access to bounds-checked field has incomplete type’

doesn't contains all the necessary information IMHO and is vague enough that I think it will cause more confusion than the current diagnostic text.

* These fields are only bounds checked with `-fbounds-safety`. Without `-fbounds-safety` the attributes are just hints for `__builtin_dynamic_object_size`. So calling the field bounds checked isn't accurate unless `-fbound-safety` is on. Even then this is vague enough that someone might not understand this is related to `-fbounds-safety`.
* It's not really obvious what "bounds-checked" field means (given we have multiple implementations of bounds checking in clang) and  more information is **neccessary** to understand that this restriction comes from the  `__counted_by` attribute . I don't think this **necessary** information should be moved into a note.

I'll grant you the diagnostic text is long but I believe almost everything there is **necessary** to understand the error. However, the `'; consider providing...` text isn't strictly necessary and could be moved to a note.

So I'd prefer to either leave the diagnostic text as it is, or move the `; consider...` part of the text to its own note.

Please let me know what you'd prefer.

AaronBallman replied:
We actually have some documentation on this: https://clang.llvm.org/docs/InternalsManual.html#diagnostic-wording -- basically, we want the wording to be succinct but actionable.

I think we could make the diagnostic a bit more succinct without losing too much information. What about something along the lines of:
```
cannot use 'explicit_desig_init.buf_typedef' with type 'Incomplete_t * __counted_by(count)' (aka 'struct IncompleteTy *') because the pointee type is incomplete; consider completing the definition or using the '__sized_by' attribute instead
```
WDYT?

delcypher replied:
This is more succinct but it looses the important piece of information that is restriction comes from the `__counted_by` attribute which is really not obvious which the current diagnostic text (tries to) convey but the proposed diagnostic text does not.

delcypher replied:
@AaronBallman Just drawing attention to this reply. For some reason it isn't showing up in the conversation tab (I think it's [this problem](https://github.com/orgs/community/discussions/6893#discussioncomment-8309254))

Sirraide replied:
> This is more succinct but it looses the important piece of information that is restriction comes from the `__counted_by` attribute which is really not obvious.

Maybe instead of `cannot use X with type ...`, we could write `cannot use X with '__counted_by'-attributed type ...`? I feel like that might be as short as we’ll be able to make it.

Also, in addition to that, maybe we can move the ‘consider completing the definition or using the '__sized_by' attribute instead’ to a note that points to the attribute location... if we have access to that here, that is (iirc there were some problems w/ that, but I’m not sure anymore)?

delcypher replied:
@Sirraide Thank you for these suggestions. I applied these in 

https://github.com/llvm/llvm-project/pull/106321/commits/0c9c62b60d7d602b87aa3bb612d1240cc70f470b

I realized while doing this that

* The existing note in my PR about where the `__counted_by` attribute could be declared could be replaced by the new diagnostic that suggests using `__sized_by` instead.
* The existing note in my PR about where the incomplete type is declared could be replaced with the new diagnostic that suggests completing type.

The means that the number of notes we emit has not actually increased between my original patch and this new one which is nice. So overall I think this is a really nice clean up so thank's for suggesting it.

AaronBallman commented on clang/lib/AST/Type.cpp:
```diff
@@ -3907,6 +3923,31 @@ CountAttributedType::CountAttributedType(
+// of the attribute that was used in the user's code. This method is used for
+// diagnostics so the fact it doesn't use the spelling of the attribute in
+// the user's code could be confusing (#113585).
+#define ENUMERATE_ATTRS(PREFIX)                                                \
```
Lambda instead of a macro?

delcypher replied:
I did this way because it avoids allocating anything because they are all constant strings. This is probably premature optimization though. I'll try to rewrite using a lambda.

delcypher replied:
Hmm actually I think I would need support for `constexpr` `std::string` to implement this and still be able to return a `StringRef`. I think LLVM is still on c++17 so I'm not sure using a lambba is do-abe here.

Sirraide replied:
Yeah, the string literal concatenation kind more or less forces this to be a macro from what I can tell, and the underlying problem is that we don’t have access to the actual attribute, but fixing that issue is a bit tricky unfortunately...

Sirraide replied:
> Hmm actually I think I would need support for `constexpr` `std::string` to implement this

(fyi, even `constexpr std::string` wouldn’t be enough; an implementation of this that uses `std::string` would require non-transient `constexpr` allocation, which is something people are working on, but it’s not in the standard yet and no compiler supports anything like it afaik)

AaronBallman commented on clang/lib/AST/Type.cpp:
```diff
@@ -3907,6 +3923,31 @@ CountAttributedType::CountAttributedType(
 }
 
+StringRef CountAttributedType::getAttributeName(bool WithMacroPrefix) const {
+// TODO: This method isn't really ideal because it doesn't return the spelling
```
It's not clear to me why this is needed at all -- doesn't `Attr::getSpelling()` suffice on the attribute, rather than asking the type?

AaronBallman commented on clang/include/clang/Sema/Sema.h:
```diff
@@ -2068,13 +2068,55 @@ class Sema final : public SemaBase {
+  /// \param E the expression to check
+  ///
+  /// \returns True iff no diagnostic where emitted, false otherwise.
+  bool BoundsSafetyCheckUseOfCountAttrPtr(Expr *E);
```
```suggestion
  bool BoundsSafetyCheckUseOfCountAttrPtr(const Expr *E);
```
Can this be made `const`? I would expect a function named "check" doesn't need to mutate the expression.

delcypher replied:
Seems to work. I've made the change.

AaronBallman commented on clang/lib/Sema/SemaBoundsSafety.cpp:
```diff
@@ -102,7 +105,36 @@ bool Sema::CheckCountedByAttrOnField(FieldDecl *FD, Expr *E, bool CountInBytes,
   // when `FieldTy->isArrayType()`.
   bool ShouldWarn = false;
-  if (PointeeTy->isIncompleteType() && !CountInBytes) {
+  if (PointeeTy->isAlwaysIncompleteType() && !CountInBytes) {
```
Perhaps silly question: did you consider a model where you track incomplete types or their checks to ensure they're completed before the end of the TU and issuing a diagnostic for any that remain incomplete at the end of the TU? We do this dance somewhat frequently: https://github.com/llvm/llvm-project/blob/0d9cf2671e06c9124a0b5fc753330c39c8b4a791/clang/lib/Sema/Sema.cpp#L1171

delcypher replied:
It's not a silly question at all. This is not a model I explored in much depth.

One reason to not do it that way is that our current implementation requires that "uses" of pointers annotated with `__counted_by` have a complete pointee type at the "use" site because in some cases we represent bounds checks at the AST level and IIRC the creation of the AST node that represents the bounds check needs to know the pointee size.

This is a design decision that could (and probably should be) revisited at some point but I don't think this is the right time to do it.

As noted in the PR description there is a case with tentative definitions that isn't handled correctly right now and would require waiting until the end of the translation unit to handle correctly. It's been omitted because it's not something we've needed so we've simply not implemented it yet.

Sirraide replied:
We could revisit that at some point, but I don’t think that needs to be part of this pr yeah.

AaronBallman commented on clang/lib/Sema/SemaBoundsSafety.cpp:
```diff
@@ -186,4 +218,206 @@ bool Sema::CheckCountedByAttrOnField(FieldDecl *FD, Expr *E, bool CountInBytes,
+    //
+    // TODO: Implement logic to find the relevant TypeLoc for the attribute and
+    // get the SourceRange from that (#113582).
+    auto AttrSrcRange = CATy->getCountExpr()->getSourceRange();
```
```suggestion
    SourceRange AttrSrcRange = CATy->getCountExpr()->getSourceRange();
```

delcypher replied:
I'm happy to make this change but why is `auto` discouraged here?

erichkeane replied:
https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable

Basically, if you can't tell the type from the RHS (or if the type is irrelevant, such as iterator types), we choose not to use 'auto'.

AaronBallman commented on clang/lib/Sema/SemaBoundsSafety.cpp:
```diff
@@ -186,4 +218,216 @@ bool Sema::CheckCountedByAttrOnField(FieldDecl *FD, Expr *E, bool CountInBytes,
+/// \param RHSExpr The expression being assigned from.
+/// \param Action The type assignment being performed
+/// \param Loc The SourceLocation to use for error diagnostics
+/// \param ComputeAssignee If provided this function will be called before
+///        emitting a diagnostic. The function should return the name of
+///        entity being assigned to or an empty string if this cannot be
+///        determined.
+///
+/// \returns True iff no diagnostic where emitted, false otherwise.
+static bool CheckAssignmentToCountAttrPtrWithIncompletePointeeTy(
+    Sema &S, QualType LHSTy, Expr *RHSExpr, AssignmentAction Action,
+    SourceLocation Loc, llvm::function_ref<std::string()> ComputeAssignee) {
```
Yeah, it doesn't seem like we need a function reference here. If we do, this could be changed to a templated function which accepts a callable, and that would still be better than using a `function_ref`.

delcypher replied:
> All the uses I saw at least end up passing in the name of some declaration—or are there cases where we need/want to pass in some other string?

* So in some cases there isn't a named assignee. This is the case for  return values as can be seen [here](https://github.com/swiftlang/llvm-project/blob/b0bbd7e8a003990aff9ce373bc6b83407630b5f5/clang/lib/Sema/SemaChecking.cpp#L9505). This is also the case for unnamed parameters as can be seen [here](https://github.com/swiftlang/llvm-project/blob/b0bbd7e8a003990aff9ce373bc6b83407630b5f5/clang/lib/Sema/SemaBoundsSafety.cpp#L640)
* In some cases we want the fully qualified name of a field in the struct. You can see this [here](https://github.com/swiftlang/llvm-project/blob/b0bbd7e8a003990aff9ce373bc6b83407630b5f5/clang/lib/Sema/SemaBoundsSafety.cpp#L537) for initializer lists and [here](https://github.com/swiftlang/llvm-project/blob/b0bbd7e8a003990aff9ce373bc6b83407630b5f5/clang/lib/Sema/SemaExpr.cpp#L16743) for assignment expressions

The reason the function takes a `function_ref` is I realized it was pretty wasteful to always allocate a `std::string` and compute it in the cases where the diagnostic is not emitted (the common case). So by making the allocation and computation of the string a `function_ref` this can be performed **only in the case that the diagnostic is emitted**.

That being said it might be possible to replace the `function_ref` by instead passing a `std::optional<clang::NamedDecl>` instead. I'll need to investigate if I can make this work correclty.

delcypher replied:
@AaronBallman 
> Yeah, it doesn't seem like we need a function reference here. If we do, this could be changed to a templated function which accepts a callable, and that would still be better than using a function_ref.

I'm not sure what you mean by that. How does making `CheckAssignmentToCountAttrPtrWithIncompletePointeeTy` a templated function help?

delcypher replied:
I tried reworking this in our fork that contains the entire `-fbounds-safety` and it seems to work: https://github.com/swiftlang/llvm-project/pull/9807/files

Once I get that integrated I'll make the same change to this PR.

Sirraide replied:
> I'm not sure what you mean by that. How does making `CheckAssignmentToCountAttrPtrWithIncompletePointeeTy` a templated function help?

Basically, instead of doing
```c++
void foo(function_ref<some_function_type> f) { ... }` // or `std::function` instead of `function_ref`
```
which performs type erasure (and may allocate in the case of `std::function`), you can write
```c++
template <typename Callable>
void foo(Callable f)
```
because then you don’t need type erasure because you can just pass in the callable directly.

> I tried reworking this in our fork that contains the entire -fbounds-safety and it seems to work: https://github.com/swiftlang/llvm-project/pull/9807/files

I see you’re passing in the `ValueDecl` instead of a `function_ref` now; that’s also fine of course if that works for all cases.

AaronBallman commented on clang/test/Sema/attr-counted-by-late-parsed-struct-ptrs.c:
```diff
@@ -19,13 +19,12 @@ struct on_member_pointer_complete_ty {
 
 struct on_member_pointer_incomplete_ty {
-  struct size_unknown * buf __counted_by(count); // expected-error{{'counted_by' cannot be applied to a pointer with pointee of unknown size because 'struct size_unknown' is an incomplete type}}
+  struct size_unknown * buf __counted_by(count); // ok
```
IMO, this seems slightly problematic for API designers, which is why I was asking about delaying checks to the end of the TU above. For someone designing an API with poor testing, they can miss the fact that this attribute is incorrect because nothing ever completes the type in this TU; I think both of these cases should still be diagnosed.

WDYT?

delcypher replied:
The current design is very much intentional.

Yes it's potentially problematic for API designers but most people are **not API designers**. For people who are not API designers why should consumers of that header need to care about an incomplete type if they never try to use it?

The current design also allows for something like:

```
// public_header.h

// Public and  implementation use same struct representation for `Handle`.
struct OpaqueData;
struct Handle {
  struct OpaqueData* __counted_by(count) Objects;
  size_t count;
 };

// library_internals.c

// Only the library internals knows how `OpaqueData` is defined.
struct OpaqueData {
  int real_data;
};

struct Handle alloc_handle(size_t num_objects) {
  struct Handle h;
  h.Objects = malloc(num_objects);
  h.count = num_objects

 if (!h.Objects)
  h.count = 0;

 return h;
}
```

That said we should try to help out API designers where possible. We could consider potentially emitting warnings (that have a sane suppression mechanism) for `__counted_by` pointers where the pointee never gets completed in a TU. However, if we really want that I think that belongs in a separate PR.

@rapidsna Do you have anything to add here?

Sirraide replied:
Yeah, that makes sense. Being able to benefit from the attribute in the implementation of a library without requiring a type to be complete at the API level seems like a reasonable use case.

delcypher commented on clang/lib/Sema/SemaBoundsSafety.cpp:
```diff
@@ -186,4 +218,216 @@ bool Sema::CheckCountedByAttrOnField(FieldDecl *FD, Expr *E, bool CountInBytes,
+  assert(CATy && !CATy->isCountInBytes() && !PointeeTy.isNull());
+
+  // It's not expected that the diagnostic be emitted in these cases.
+  // It's not necessarily a problem but we should catch when this starts
+  // to happen.
```
I think when I wrote this comment I was thinking of this in the context of test coverage. I'll just drop the assert and comments.

delcypher commented on clang/lib/Sema/SemaBoundsSafety.cpp:
```diff
@@ -186,4 +218,216 @@ bool Sema::CheckCountedByAttrOnField(FieldDecl *FD, Expr *E, bool CountInBytes,
+
+      // It's not necessarily bad if this assert fails but we should catch
+      // if this happens.
+      assert(Entity.getKind() == InitializedEntity::EK_Member);
```
I've fixed both problems.

delcypher commented on clang/lib/Sema/SemaBoundsSafety.cpp:
```diff
@@ -186,4 +218,216 @@ bool Sema::CheckCountedByAttrOnField(FieldDecl *FD, Expr *E, bool CountInBytes,
+  if (!HasCountedByAttrOnIncompletePointee(T, &IncompleteTyDecl, &CATy,
+                                           &PointeeTy))
+    return true;
+  assert(CATy && !CATy->isCountInBytes() && !PointeeTy.isNull());
```
I removed the asserts entirely because if you move them into the implementation its pretty obvious they cannot fail.

delcypher commented:
@Sirraide Thanks for your review and sorry for the delay in getting back to this. I've tried to address as much as possible and left comment on things that still need discussion. Please take a look when time permits.

delcypher commented:
@erichkeane @AaronBallman @Sirraide I've addressed everything I can. 

I seem to be running into a problem where some of my replies to comments are not showing up in the conversation tab. This seems to be a [known problem with GitHub reviews ](https://github.com/orgs/community/discussions/6893#discussioncomment-8309254):( . You may have to dig into the Files tab to see some of my replies to comments :(

delcypher commented:
[This comment](https://github.com/llvm/llvm-project/pull/106321/files#r1906005324) is not showing up in the conversation view for some reason.

Sirraide commented:
> I seem to be running into a problem where some of my replies to comments are not showing up in the conversation tab.

Yeah, I prefer looking at the flood of emails I get from GitHub, because those pretty consistently *do* include every comment.

Sirraide commented on clang/include/clang/Sema/Sema.h:
```diff
@@ -2068,13 +2068,57 @@ class Sema final : public SemaBase {
   // -------------------------------------------------------------------------
   //
-  //
+  //  
```
```suggestion
  //
```
Whitespace at the end of the line

Sirraide commented on clang/lib/Sema/SemaBoundsSafety.cpp:
```diff
@@ -186,4 +218,188 @@ bool Sema::CheckCountedByAttrOnField(FieldDecl *FD, Expr *E, bool CountInBytes,
+  if (CATy->isCountInBytes())
+    return {};
+
+  auto PointeeTy = CATy->getPointeeType();
+  if (PointeeTy.isNull())
+    return {}; // Reachable?
```
```suggestion
  auto PointeeTy = CATy->getPointeeType();
  if (PointeeTy.isNull())
    return {}; // Reachable?
```
Does a `CountAttributedType` *always* desugar to a `PointerType`? If so, then I don’t think this is reachable because `getPointeeType()` ignores type sugar. In that case, we might want to add `assert(!PointeeTy.isNull())` here?

delcypher replied:
Not always. For a flexible array member like this.

```
struct FAM {
  int size;
  char buffer[] __counted_by(size);
};
```

The type of the `buffer` member is

```
CountAttributedType 0x121079600 'char[] __counted_by(count)' sugar
`-IncompleteArrayType 0x121079500 'char[]' 
  `-BuiltinType 0x121018cb0 'char'
```

so that desugars to an `IncompleteArrayType` rather than a `PointerType`.

I'm not sure if this path is reachable in the contexts that `HasCountedByAttrOnIncompletePointee` is currently called on. It is an error to try to make a flexible-array-member with an incomplete type.

```
struct IncompleteTy;
struct FAM {
  int size;
  struct IncompleteTy buffer[] __counted_by(size); // error
};
```

`HasCountedByAttrOnIncompletePointee` wasn't reachable when I tried simple examples.

That being said we may want to call `HasCountedByAttrOnIncompletePointee` in different contexts in the future so I think returning rather than asserting is the right thing to do here.

delcypher replied:
I left a comment in the code about the case where the early return is reachable.

Sirraide commented on clang/lib/Sema/SemaBoundsSafety.cpp:
```diff
@@ -186,4 +218,188 @@ bool Sema::CheckCountedByAttrOnField(FieldDecl *FD, Expr *E, bool CountInBytes,
+
+static std::tuple<const CountAttributedType *, QualType>
+HasCountedByAttrOnIncompletePointee(QualType Ty, NamedDecl **ND) {
+  auto *CATy = Ty->getAs<CountAttributedType>();
+  if (!CATy)
+    return {};
+
+  // Incomplete pointee type is only a problem for
+  // counted_by/counted_by_or_null
+  if (CATy->isCountInBytes())
+    return {};
```
```suggestion
  // Incomplete pointee type is only a problem for
  // counted_by/counted_by_or_null.
  auto *CATy = Ty->getAs<CountAttributedType>();
  if (!CATy || CATy->isCountInBytes())
    return {};
```
This is a bit terser and imo still as clear.

Sirraide commented on clang/lib/Sema/SemaBoundsSafety.cpp:
```diff
@@ -186,4 +218,188 @@ bool Sema::CheckCountedByAttrOnField(FieldDecl *FD, Expr *E, bool CountInBytes,
+    }
+  }
+  {
+    auto D = S.Diag(Loc, diag::err_counted_by_on_incomplete_type_on_assign)
+             << static_cast<int>(Action) << AssigneeStr
+             << (AssigneeStr.size() > 0) << isa<ImplicitValueInitExpr>(RHSExpr)
+             << LHSTy << CATy->getAttributeName(/*WithMacroPrefix=*/true)
+             << PointeeTy << CATy->isOrNull();
+
+    if (RHSExpr->getSourceRange().isValid())
+      D << RHSExpr->getSourceRange();
+  }
```
Have you run into any concrete problems w/ the source range being invalid here? *Generally*, we expect the vast majority of source ranges to be valid, so unconditionally passing one to a diagnostic shouldn’t be an issue.

delcypher replied:
I've not run into issues. This was just me being very defensive. I assumed attaching an invalid source range would trigger a crash/assert.

I'll remove the guard.

delcypher replied:
Hmm actually wait... I'm not sure the source range is valid when RHSExpr is an `ImplicitValueInitExpr` which it is in some cases.

delcypher replied:
This seems to work despite the invalid source range. It looks like `TextDiagnostic::emitDiagnosticLoc` ignores invalid source ranges.

delcypher replied:
And `mapDiagnosticRanges` in `clang/lib/Frontend/DiagnosticRenderer.cpp` (used in `DiagnosticRenderer::emitCaret`) also ignores the invalid source range. So at a glance unconditionally passing invalid source ranges seems to not break things. So I will remove the guard.

Sirraide commented on clang/lib/Sema/SemaBoundsSafety.cpp:
```diff
@@ -186,4 +218,188 @@ bool Sema::CheckCountedByAttrOnField(FieldDecl *FD, Expr *E, bool CountInBytes,
+
+  NamedDecl *IncompleteTyDecl = nullptr;
+  auto [CATy, PointeeTy] =
+      HasCountedByAttrOnIncompletePointee(T, &IncompleteTyDecl);
```
```suggestion
      HasCountedByAttrOnIncompletePointee(T, &IncompleteTyDecl);
```
I wonder if this function shouldn’t be renamed to something like `GetCountedByAttrOnIncompletePointee` at this point because it’s a bit weird to start a function w/ `Has` now that it doesn’t return a `bool` anymore imo.

delcypher replied:
Agreed. I'll rename.

Sirraide reviewed (commented):
Alright, there are a few minor things I pointed out here, but nothing major. I think we should be able to land this soon barring any unforeseen complications. ;Þ

(I also replied to a few comment threads, but that was before I did the actual review so it isn’t shown here unfortunately...)

delcypher commented:
@Sirraide Apologies for taking so long to get back to this PR. I think I've addressed all your feedback now. Is this ok to land now?

Sirraide reviewed (approved):
I think this should be fine at this point. I’m having a bit of a hard time trying to figure out what changed since my last review because of the force-pushing, but so long as you haven’t made any other major changes since then then this ltgm now.

delcypher commented:
> I think this should be fine at this point. I’m having a bit of a hard time trying to figure out what changed since my last review because of the force-pushing, but so long as you haven’t made any other major changes since then then this ltgm now.

Sorry about that. This PR had become so stale that there merge conflicts and the easiest way to fix it was to squash because the conflicts were with things I later removed.

Thanks for reviewing and approving. I'm going to cherry-pick this into the Swiftlang fork of llvm-project first and fix everything up (there's stuff that wasn't upstreamed that's going to conflict with this PR). Once I've done that I'll merge this.

Sirraide commented:
> Sorry about that. This PR had become so stale that there merge conflicts and the easiest way to fix it was to squash because the conflicts were with things I later removed.

I see, yeah, it happens; that’s unfortunate...

delcypher committed:
Diagnostic fix ups

delcypher commented:
I had to tweak some minor issues with the diagnostic text that I discovered when trying to cherry pick this into our fork. I've fixed them and I'm going to land this now.
